### PR TITLE
187216522 international schools

### PR DESCRIPTION
**Pull request by ArushC**
### [Pivotal Tracker Link][tracker]

[tracker]: https://www.pivotaltracker.com/story/show/your-story-id

## What this PR does:

This is a duplicate of [https://github.com/cs169/BJC-Teacher-Tracker-App/pull/42](https://github.com/cs169/BJC-Teacher-Tracker-App/pull/42). For ease's sake, do NOT review this PR until first reviewing [https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/294](https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/294) and [https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/293](https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/293). Due to an unfortunate merge conflict and a glitch with Github's web interface, we had no choice but to merge the entire main branch into the international schools branch. The main branch includes the features whose corresponding PRs were listed above.

"This PR adds the country field to the school model. Whenever a new school is created, the option to select a country is available. Moreover, the country data is introduced in tables, teacher and school information panels and in the backend. The UI logic either renders a textfield if the country selected when creating a school is not the US or a dropdown of US states if the selected country is the US."

This pull request fixes|implements (pick one...) ______.

#### Who authored this PR?

@razztech -  overall implementation, testing, code refactoring and UI logic
@perryzjc -  pair-programming help for debugging teacher requests table
@ArushC - bug fixes, additional test cases, PR fixes

### How should this PR be tested?

* Is there a deploy we can view?
* What do the specs/features test?
* Are there edge cases to watch out for?

### Checklist:

- [x] Has this been deployed to a staging environment or reviewed by a customer?
- [x] Tag someone for code review (either a coach / team member)
- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g. `michael/12345-add-new-feature` Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`)

### CODE REVIEW

**Review by fp456 — COMMENTED** [review]
Good job! Just some minor things

**Review by fp456 — APPROVED** [review]

File: app/models/email_template.rb [hunk 1]
```diff
@@ -22,8 +22,8 @@ class EmailTemplate < ApplicationRecord
   validates :title,
             inclusion: TeacherMailer.instance_methods(false).map { |method| method.to_s.titlecase },
             if: -> { self.required? }
-  validates :body, presence: { message: "cannot be blank" }
-  validates :to, presence: { message: "cannot be blank" }
+  validates :body, presence: true
```
Comment (cycomachead) [inline app/models/email_template.rb:25]: Nothing wrong with these changes, but they probably belong in a separate PR

File: app/controllers/schools_controller.rb [hunk 2]
```diff
@@ -12,11 +12,11 @@ def show
   end
 
   def search
-    School.all.collect { |school| ["#{school.name}, #{school.city}, #{school.state}", school.name] }
+    School.all.collect { |school| ["#{school.name}, #{school.country}, #{school.city}, #{school.state}", school.name] }
```
Comment (cycomachead) [inline app/controllers/schools_controller.rb:15]: Hmm. OK, I think we might want a different method here. 

The presentation should definitely be:
"School Name, City, State, Country" if we have all the data.
I wonder if the format is actually best as:
"School Name (City, State)" if the country is USA
and 
"School Name (City, State, Country)" when the country is not the US.

That logic is a little complex, so it might make sense to write a different method...

```
     def format_school([name, city, state, country]):
        country_str = country == "US" ? '' : ", #{country}"
        "#{name} (#{city}, #{state}#{country_str})"
    end

    # This should be a model method, perhaps called self.search_list on the School model.
    School.all.pluck(:name, :city, :state, :country).map { |data| [ format_state(data), name ] }
```
Use `pluck` here isn't strictly necessary, but it prevents Rails from creating 1000's of additional objects which uses a lot less memory.


```suggestion
    School.search_list
```

File: app/controllers/schools_controller.rb [hunk 3]
```diff
@@ -12,11 +12,11 @@ def show
...
   end
 
   def create
-    @school = School.find_by(name: school_params[:name], city: school_params[:city], state: school_params[:state])
+    @school = School.find_by(name: school_params[:name], country: school_params[:country], city: school_params[:city], state: school_params[:state])
```
Comment (cycomachead) [inline app/controllers/schools_controller.rb:19]: This is starting to look / read a little unwieldy.

I might write a method called `unique_school_params` or something, which then just permit's only the valid list of parameters. (You can have multiple method which do parameter filtering)

```suggestion
    @school = School.find_by(unique_school_params)
```

File: app/controllers/schools_controller.rb [hunk 4]
```diff
@@ -44,11 +44,13 @@ def edit
   def update
     @school = School.find(params[:id])
     @school.assign_attributes(school_params)
+    debugger
```
Comment (cycomachead) [inline app/controllers/schools_controller.rb:47]: This needs to be removed. Not sure if this was causing tests to fail, but if not then some tests are missing.

File: app/controllers/teachers_controller.rb [hunk 5]
```diff
@@ -100,7 +100,10 @@ def update
       @teacher.school = @school
     else
       @school.update(school_params) if school_params
-      @school.save!
+      unless @school.save
+        flash[:alert] = "An error occurred: #{@school.errors.full_messages.join(', ')}"
+        render "new" && return
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:105]:
```suggestion
        render "edit" && return
```

Or am i missing something?

File: app/controllers/teachers_controller.rb [hunk 6]
```diff
@@ -202,7 +205,7 @@ def load_school
     if teacher_params[:school_id].present?
       @school ||= School.find(teacher_params[:school_id])
     end
-    @school ||= School.find_or_create_by(name: school_params[:name], city: school_params[:city], state: school_params[:state])
+    @school ||= School.find_or_create_by(name: school_params[:name], city: school_params[:city], country: school_params[:country], state: school_params[:state])
```
Comment (cycomachead) [inline app/controllers/teachers_controller.rb:208]: This would we another great case for a `unique_school_params` method!

File: app/models/school.rb [hunk 7]
```diff
@@ -25,10 +26,12 @@
 #
 
 class School < ApplicationRecord
-  VALID_STATES = [ "AL", "AK", "AS", "AZ", "AR", "CA", "CO", "CT", "DE", "DC", "FM", "FL", "GA", "GU", "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MH", "MD", "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ", "NM", "NY", "NC", "ND", "MP", "OH", "OK", "OR", "PW", "PA", "PR", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VI", "VA", "WA", "WV", "WI", "WY", "International"].freeze
+  VALID_STATES = [ "AL", "AK", "AS", "AZ", "AR", "CA", "CO", "CT", "DE", "DC", "FM", "FL", "GA", "GU", "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MH", "MD", "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ", "NM", "NY", "NC", "ND", "MP", "OH", "OK", "OR", "PW", "PA", "PR", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VI", "VA", "WA", "WV", "WI", "WY"].freeze
```
Comment (cycomachead) [inline app/models/school.rb:29]: Ah heck, since we have set up the countries gem, let's actually just kill this list.

```
  VALID_STATES = ISO... .subdivisions.keys.freeze
```
I think should do the trick

https://github.com/countries/countries?tab=readme-ov-file#subdivisions--states

File: app/views/schools/_form.html.erb [hunk 8]
```diff
@@ -1,26 +1,37 @@
 <div class="form-group h3">Create a new School</div>
 <div class='form-group row'>
-  <div class='col-6'>
+  <div class='col'>
       <%= f.label :name, "School Name", class: "label-required" %>
-      <%= f.text_field :name, placeholder: 'UC Berkeley', class: 'form-control',
+      <%= f.text_field :name, placeholder: 'Name of School', class: 'form-control',
```
Comment (cycomachead) [inline app/views/schools/_form.html.erb:5]: Let's revert these back to what they were. 

In general, a `placeholder` value should be an example of the data, and the label should be the description.

File: app/views/schools/_form.html.erb [hunk 9]
```diff
@@ -88,3 +99,35 @@
...
+    $(document).ready(function() {
+        const countrySelected = $('#school_country');
+        const stateSelectContainer = $('#state_select_container');
+        const stateTextfieldContainer = $('#state_textfield_container');
+        const stateSelect = $('#state_select');
+        const stateTextfield = $('#state_textfield');
+
+        function handleCountryChange() {
+            if (countrySelected.val() === 'US') {
+                stateSelectContainer.show();
```
Comment (cycomachead) [inline app/views/schools/_form.html.erb:113]: The ordering/grouping of these lines this a little confusing.

File: app/views/schools/_form.html.erb [hunk 10]
```diff
@@ -88,3 +99,35 @@
...
+                stateSelect.attr('required', '')  //make state select required
+                stateTextfieldContainer.hide();
+
+                stateTextfield.removeAttr('name');
+                stateSelect.attr('name', 'school[state]');
+
```
Comment (cycomachead) [inline app/views/schools/_form.html.erb:119]:
```suggestion
                stateSelectContainer.show();
                stateSelect.attr('name', 'school[state]');.attr('required', '')  // make state select required
                
                stateTextfieldContainer.hide();
                stateTextfield.removeAttr('name');
```
If there's going to be a newline, then the two types should probably be grouped together.

File: app/views/teachers/_teacher.erb [hunk 11]
```diff
@@ -27,6 +27,9 @@
   <%= link_to(teacher.school.name, school_path(teacher.school)) %>
   <%= teacher.school.location %>
 </td>
+<td>
+  <%= teacher.school.country %>
+</td>
```
Comment (cycomachead) [inline app/views/teachers/_teacher.erb:32]: Actually let's remove this from the teacher's table.

File: app/views/teachers/_table_headers.erb [hunk 12]
```diff
@@ -4,5 +4,6 @@
 <th scope="col">Status</th>
 <th scope="col" width="15ch">Snap<i>!</i></th>
 <th scope="col">School</th>
+<th scope="col">Country</th>
```
Comment (cycomachead) [inline app/views/teachers/_table_headers.erb:7]: remove this line.

File: app/views/teachers/_teacher.erb [hunk 13]
```diff
@@ -27,6 +27,9 @@
   <%= link_to(teacher.school.name, school_path(teacher.school)) %>
   <%= teacher.school.location %>
```
Comment (cycomachead) [inline app/views/teachers/_teacher.erb:28]: Let's make sure the location method is updated to show the country.

File: app/models/school.rb
Comment (cycomachead) [inline app/models/school.rb:1]: OK yeah, we should update the `location` and `name_location` methods to show the country.

I think, `location` should include the country text (city, state, country) format if the country != 'US'. Otherwise, `name_location` should always include the country since it passes the result to the google maps API

Do we need to update the `selectize` methods in this file?

**Review by cycomachead — CHANGES_REQUESTED** [review]
Looks very good! Some mostly minor comments, but a couple things to refactor. 

* UI: We should simplify the location / display info of schools because the tables get large
* we should refactor some controller code, because it's getting a bit complex.

**Review by ArushC — COMMENTED** [review]
Line 21 here needs to be kept for the cucumber tests to pass. This conflict was resolved incorrectly. I will try to fix it right now

File: app/models/school.rb
Comment (ArushC) [reply app/models/school.rb:1]: Yes, I think so.

**Comment by ArushC** [comment]
@cycomachead I made all the changes that you requested except removing the countries thing from the table because that resulted in a really weird bug which caused some Cucumber tests related to the admin feature to unexpectedly fail. I could look into it more, but my 10 min glance says it would probably be a 2-3 point story to figure out what's going on there. 

Otherwise, I think this PR is in good shape. Feel free to take a look at the stuff that I changed and merge this.

**Review by cycomachead — CHANGES_REQUESTED** [review]
Sorry, I had some comments open for a while, but there are few things to fix.

File: app/models/school.rb [hunk 14]
```diff
@@ -98,6 +101,18 @@ def maps_marker_data
...
+  def format_school(data)
+    name, city, state, country = data
+    country_str = country == "US" ? "" : ", #{country}"
+    "#{name} (#{city}, #{state}#{country_str})"
+  end
+
+  def self.search_list
+    School.pluck(:name, :city, :state, :country).map do |data|
+      format_school(data)
+    end
```
Comment (cycomachead) [inline app/models/school.rb:113]:
```suggestion
    School.pluck(:name, :city, :state, :country).map(&:format_school)
```
